### DISCUSSION TITLE
Write Float Values on Modbus Holding Register Address

### DISCUSSION THREAD

**Discussion by Ricdo7** [pymodbus-dev/pymodbus#1458] [Help] [answered]
I know that modbus does not support floating, DINT and 32bits. I know that you can send a float and that it is possible, but that it is a bit tricky (I can see it myself), but in my project I have to send a float and read a float. I looked at the documentation and all kinds of videos and asked chatGPT, but now I'm slowly getting lost and writing nonsense.

I just want to write 7.7 to PLC memory %MW1 (Real) and read from another register 3.3.

I would be very grateful if you could give me a concrete answer, and not just 'read the documentation'.

Please
Thank you

#Proba
def unos():
    str_vrijednost=(float(ulazna_velicina.get()))
    print(str_vrijednost)
    print(type(str_vrijednost))
    
    builder = BinaryPayloadBuilder(byteorder= Endian.Big, wordorder=Endian.Little)
    builder.add_32bit_float(str_vrijednost)
    dint_pisi = builder.build()
    
    payload_bytes = bytearray(dint_pisi)
    payload_int = int.from_bytes(payload_bytes, byteorder='big',signed=True)
    
    client.write_register(4,payload_int,skip_encode=True)

**Comment by janiversen** [comment]
The payload builder can do exactly what you need, however if you look at dev (as in not yet released) I have added convert_to_register and convert_from_register functions, that makes life very easy for you.

**Reply by Ricdo7** [reply]
Jan thanks for your reply,
I didn't really understand you. If I understood correctly, you will put an example or something similar about (convert_from_register functions and convert_to_register ) in pymodbus-dev. When could I expect this in dev?
Thank you very much

**Reply by janiversen** [reply]
No I wrote it is already on dev, but not in a formal release.

**Reply by janiversen** [reply]
It will be part of version 3.3 due around June this year.

**Reply by Ricdo7** [reply]
> No I wrote it is already on dev, but not in a formal release.

Can you give me link to this,please?
When i search in pymodbus-dev for 'convert_from_register functions' and 'convert_to_register' i only see my forum.

**Reply by janiversen** [reply]
A link to what? dev is a branch you can use, if you want the specific commit then look in the commits.

**Reply by Ricdo7** [reply]
I can't find where you wrote the convert_from_register functions and convert_to_register . I'm new to Github.

**Reply by janiversen** [reply]
use branch dev instead  of whatever version you are using, then  it is included.

**Reply by Ricdo7** [reply]
I'm not sure how could I write the code. If you can give me a little longer explanation or some banal example how this converter should work and write to the registry. I would be very grateful, thank you.
##########################################################################  
class DATATYPE(Enum):
    """Data type enum (name and number of bytes), used for convert_* calls."""
    BIT = 0
    UINT16 = 1
    INT16 = 2
    UINT32 = 3
    INT32 = 4
    FLOAT32 = 5
    FLOAT64 = 6
##########################################################################

#Proba
def unos():
    str_vrijednost=(float(ulazna_velicina.get()))
    print(str_vrijednost)
    print(type(str_vrijednost))
    
    payload_int=client.convert_to_registers(str_vrijednost,DATATYPE.INT16)

    client.write_register(4,payload_int)

**Reply by janiversen** [reply, marked as answer]
If you want to convert a float then use the appropriate datatype not INT16. set the number of registers in the call, and also set slave=.
